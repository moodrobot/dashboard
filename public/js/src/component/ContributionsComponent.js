import React, { PureComponent } from "react";
import PropTypes from "prop-types";
//import shallowCompare from "react-addons-shallow-compare";

//import i18n from "../i18n"; // initialized i18next instance

class ContributionsComponent extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      style: "category",
      data: [],
      filter_data: [],
      loader_display: "",
      view_more_display: "d-none",
      content_display: "d-none",
      no_content_display: "d-none",
      max: 10,
      display_num: 1
    };

    //console.log(this.props);
    this.filterArray = this.filterArray.bind(this);
    this.createContributions = this.createContributions.bind(this);
    this.handleFilter = this.handleFilter.bind(this);
    this.showMoreItems = this.showMoreItems.bind(this);
  }

  createContributions() {
    const T = this.props.translate;
    const currentLanguage = this.props.currentLanguage;
    let filter_data = [...this.state.filter_data];
    if (filter_data.length === 0) {
      filter_data = [...this.state.data];
    }
    //console.log("FILTER DATA", filter_data);
    const listItems = filter_data
      .slice(0, this.state.max * this.state.display_num)
      .map(function(item, index) {
        let link = item.linkText;
        let publishState = "";

        if (item.unpublished) {
          publishState = T["event-unpublish"];
        }

        if (item.linkUrl) {
          link = <a href={item.linkUrl}>{item.linkText}</a>;
        }

        let dateOptions = {
          year: "numeric",
          month: "numeric",
          day: "numeric",
          hour: "numeric",
          minute: "numeric",
          second: "numeric"
          // hour12: false
        };

        return (
          <tr className="contribution" key={index} value={item.value}>
            <td>
              {T["event-names"][item.nice_event_name]} {publishState}
            </td>
            <td>{link}</td>
            <td>
              {new Intl.DateTimeFormat(currentLanguage, dateOptions).format(new Date(item.awarded))}
            </td>
            <td>{item.points}</td>
          </tr>
        );
      });

    return listItems;
  }

  showMoreItems() {
    var count = this.state.display_num;
    count++;

    //set new display amount and re-render
    this.setState({ display_num: count });
  }

  static getDerivedStateFromProps(nextProps, state) {
    //console.log(nextProps.init);
    if (nextProps.init) {
      //view more display
      let vm = "";
      let cc = "d-block";
      let no_cont = "d-none";
      //console.log(nextProps.data.length, state.max);
      if (nextProps.data.length < state.max) {
        vm = "d-none";
      }

      if (nextProps.data.length == 0) {
        no_cont = "d-block";
        cc = "d-none";
      }

      if (state.display_num * state.max >= state.filter_data.length && state.filter_data.length > 0) {
        vm = "d-none";
      } else if (state.filter_data.length >= state.max) {
        vm = "";
      }

      return {
        data: nextProps.data,
        filter_data: state.filter_data,
        loader_display: "d-none",
        content_display: cc,
        view_more_display: vm,
        no_content_display: no_cont
      };
    } else {
      return null;
    }
  }

  /*componentDidUpdate() {
    let vm;

    if (this.state.display_num * this.state.max >= this.state.filter_data.length && this.state.filter_data.length > 0) {
      vm = "d-none";
    } else if (this.state.filter_data.length >= this.state.max) {
      vm = "";
    }

    //MAKE SURE THERE IS A CHANGE OR WILL INFINITE LOOP
    if (vm != this.state.view_more_display) {
      this.setState({ view_more_display: vm });
    }
  }*/

  initFilterDropDown() {
    const filter_options = [...new Set(this.state.data.map(item => item.nice_event_name))].map(function(item, index) {
      return (
        <option key={index} value={item}>
          {item}
        </option>
      );
    });

    return filter_options;
  }

  handleFilter(e) {
    this.filterArray(e.target.value);
  }

  filterArray(filter_str) {
    const temp = [...this.state.data.filter(item => item.nice_event_name === filter_str)];
    this.setState({ filter_data: temp, display_num: 1 });
  }

  render() {
    const T = this.props.translate;
    //console.log("RENDERING CONTRIBUTIONS COMPONENT");
    return (
      <div className="container-fluid">
        <div id="contribution-content" className={this.state.content_display}>
          <select
            id="select_filter"
            defaultValue="default"
            className="capitalize filter_dropdown form-control"
            onChange={this.handleFilter}
          >
            <option key="filter" disabled value="default">
              {T["filter-by"]}
            </option>
            {this.initFilterDropDown()}
          </select>
          <table className="table table-responsive">
            <thead>
              <tr>
                <th>{T["type"]}</th>
                <th>{T["title"]}</th>
                <th>{T["posted"]}</th>
                <th>{T["points-earned"]}</th>
              </tr>
            </thead>
            <tbody>{this.createContributions()}</tbody>
          </table>
        </div>
        <div className="col-12 text-center">
          <div className="loader row">
            <div id="no_cont" className={this.state.no_content_display}>
              No contributions yet
            </div>
            <div id="loader_cont" className={"col-12 " + this.state.loader_display}>
              <img src="images/loader.gif" />
            </div>
          </div>
          <button
            id="view_more_btn"
            onClick={this.showMoreItems}
            type="button"
            className={"btn btn-secondary " + this.state.view_more_display}
          >
            {this.props.view_more}
          </button>
        </div>
      </div>
    );
  }
}

ContributionsComponent.propTypes = {
  translate: PropTypes.object,
  currentLanguage: PropTypes.string,
  data: PropTypes.array,
  view_more: PropTypes.string,
  init: PropTypes.bool
};

export default ContributionsComponent;
