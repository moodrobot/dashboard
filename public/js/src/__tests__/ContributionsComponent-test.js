import React from "react";
import jest from "jest";
import { shallow, mount } from "enzyme";
import sinon from "sinon";
import ContributionsComponent from "../component/ContributionsComponent";
import contributionsData from "./__mockData__/contributionsData.json";
// import translateData from "../../../../locales/en-US.json";
import i18n from "../i18n-for-testing"; // initialized i18next instance for testing

const cc = sinon.spy(ContributionsComponent.prototype, "createContributions");

describe("<ContributionsComponent />", () => {
  const wrapper = shallow(
    <ContributionsComponent
      data={contributionsData}
      currentLanguage={"en-US"}
      translate={i18n.t("contributions")}
      view_more = {i18n.t("view-more-button")}
      init={true}
    />

  );
  it("expect to match snapshot", () => {

    //TRIGGERS COMPONENT DID UPDATE
    wrapper.setProps({data: contributionsData});

    /**EXPECT NO CHANGE IN THE SNAPSHOT */
    expect(wrapper).toMatchSnapshot();

  });
  it("expect wrapper to increment display number by 1", () => {

    /**CLICK THE VIEW MORE BUTTON */
    wrapper.find("button").first().simulate("click");

    wrapper.update();

    /**EXPECT THE DISPLAY NUM TO INCREASE BY 1 */
    expect(wrapper.state().display_num).toBe(2);

  });
  it("expect to filter by blog keyword", () => {

    /*setTimeout(function(){
      expect(wrapper.state().filter_data.length > 1);
      console.log("FILTER _DATAS", wrapper.state().filter_data.length);
      done();
    }, 0);*/

    //SIMULATE SELECTING BLOG FROM DROPDOWN
    //wrapper.find("select").simulate("change", {target: {value : "blog"}});

    //wrapper.update();

    //nexpect(wrapper.find("tbody").find("tr").length).toBe(2);

  });
});
