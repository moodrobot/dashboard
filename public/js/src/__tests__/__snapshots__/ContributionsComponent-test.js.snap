// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContributionsComponent /> expect to match snapshot 1`] = `
<div
  className="container-fluid"
>
  <div
    className="d-block"
    id="contribution-content"
  >
    <select
      className="capitalize filter_dropdown form-control"
      defaultValue="default"
      id="select_filter"
      onChange={[Function]}
    >
      <option
        disabled={true}
        key="filter"
        value="default"
      >
        Filter By...
      </option>
      <option
        key="0"
        value="admin"
      >
        admin
      </option>
      <option
        key="1"
        value="blog"
      >
        blog
      </option>
      <option
        key="2"
        value="comment"
      >
        comment
      </option>
      <option
        key="3"
        value="forum"
      >
        forum
      </option>
      <option
        key="4"
        value="article"
      >
        article
      </option>
    </select>
    <table
      className="table table-responsive"
    >
      <thead>
        <tr>
          <th>
            Type
          </th>
          <th>
            Title
          </th>
          <th>
            Posted
          </th>
          <th>
            Points Earned
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="contribution"
          key="0"
        >
          <td>
            Admin
             
          </td>
          <td>
            belt testing
          </td>
          <td>
            10/24/2017, 2:08:24 PM
          </td>
          <td>
            -1100
          </td>
        </tr>
        <tr
          className="contribution"
          key="1"
        >
          <td>
            Blog
             
            (unpublished)
          </td>
          <td>
            blog: new event name test
          </td>
          <td>
            10/13/2017, 2:36:08 PM
          </td>
          <td>
            -80
          </td>
        </tr>
        <tr
          className="contribution"
          key="2"
        >
          <td>
            Comment
             
            (unpublished)
          </td>
          <td>
            article comment #2
          </td>
          <td>
            10/13/2017, 2:07:45 PM
          </td>
          <td>
            -5
          </td>
        </tr>
        <tr
          className="contribution"
          key="3"
        >
          <td>
            Comment
             
          </td>
          <td>
            <a
              href="https://int2-software.intel.com/comment/1913417"
            >
              article comment #3
            </a>
          </td>
          <td>
            10/13/2017, 2:07:19 PM
          </td>
          <td>
            5
          </td>
        </tr>
        <tr
          className="contribution"
          key="4"
        >
          <td>
            Comment
             
          </td>
          <td>
            article comment #2
          </td>
          <td>
            10/13/2017, 2:05:04 PM
          </td>
          <td>
            5
          </td>
        </tr>
        <tr
          className="contribution"
          key="5"
        >
          <td>
            Comment
             
          </td>
          <td>
            <a
              href="https://int2-software.intel.com/comment/1913415"
            >
              blog comment #2
            </a>
          </td>
          <td>
            10/13/2017, 2:04:54 PM
          </td>
          <td>
            5
          </td>
        </tr>
        <tr
          className="contribution"
          key="6"
        >
          <td>
            Comment
             
          </td>
          <td>
            <a
              href="https://int2-software.intel.com/comment/1913414"
            >
              forum comment
            </a>
          </td>
          <td>
            10/13/2017, 2:04:07 PM
          </td>
          <td>
            20
          </td>
        </tr>
        <tr
          className="contribution"
          key="7"
        >
          <td>
            Forum
             
          </td>
          <td>
            <a
              href="https://int2-software.intel.com/node/746640"
            >
              forum: new event name test
            </a>
          </td>
          <td>
            10/13/2017, 11:27:20 AM
          </td>
          <td>
            10
          </td>
        </tr>
        <tr
          className="contribution"
          key="8"
        >
          <td>
            Blog
             
          </td>
          <td>
            blog: new event name test
          </td>
          <td>
            10/13/2017, 11:26:00 AM
          </td>
          <td>
            80
          </td>
        </tr>
        <tr
          className="contribution"
          key="9"
        >
          <td>
            Article
             
          </td>
          <td>
            <a
              href="https://int2-software.intel.com/node/746638"
            >
              article: new event name test
            </a>
          </td>
          <td>
            10/13/2017, 11:24:57 AM
          </td>
          <td>
            80
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12 text-center"
  >
    <div
      className="loader row"
    >
      <div
        className="d-none"
        id="no_cont"
      >
        No contributions yet
      </div>
      <div
        className="col-12 d-none"
        id="loader_cont"
      >
        <img
          src="images/loader.gif"
        />
      </div>
    </div>
    <button
      className="btn btn-secondary "
      id="view_more_btn"
      onClick={[Function]}
      type="button"
    >
      View More
    </button>
  </div>
</div>
`;
